nvflare/app_common/resource_managers/list_resource_manager.py

```diff
@@ -12,66 +12,142 @@
+                            self.resources[k].append(i)
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.SYSTEM_START:
+            self.cleanup_thread.start()
+        elif event_type == EventType.SYSTEM_END:
+            self.stop_event.set()
+            if self.cleanup_thread:
+                self.cleanup_thread.join()
+                self.cleanup_thread = None
+
+    def _check_all_required_resource_available(self, resource_requirement: dict, fl_ctx: FLContext) -> bool:
         check_result = True
-        for k in resource_requirement:
-            if k in self.resources:
-                if len(self.resources[k]) < resource_requirement[k]:
+        with self.lock:
+            for k in resource_requirement:
+                if k in self.resources:
+                    if len(self.resources[k]) < resource_requirement[k]:
+                        check_result = False
+                        self.log_debug(fl_ctx, f"Resource {k} is not enough.")
+                        break
+                else:
                     check_result = False
+                    self.log_debug(fl_ctx, f"Missing {k} in resources.")
                     break
-        return check_result, str(uuid.uuid4())
+        return check_result
+
+    def check_resources(self, resource_requirement: dict, fl_ctx: FLContext) -> (bool, Optional[str]):
+        if not isinstance(resource_requirement, dict):
+            raise TypeError(f"resource_requirement should be of type dict, but got {type(resource_requirement)}.")
+
+        check_result = self._check_all_required_resource_available(resource_requirement, fl_ctx)
+        token = None
+
+        # reserve resource only when check is True
+        if check_result:
+            token = str(uuid.uuid4())
+            reserved_resources = {}
+            with self.lock:
+                for k in resource_requirement:
+                    reserved_resource_units = []
+                    for i in range(resource_requirement[k]):
+                        reserved_resource_units.append(self.resources[k].popleft())
+                    reserved_resources[k] = reserved_resource_units
+                self.reserved_resources[token] = (reserved_resources, self.expiration_period)
+        return check_result, token
 
     def cancel_resources(self, resource_requirement: dict, token: str, fl_ctx: FLContext):
+        with self.lock:
+            if token and token in self.reserved_resources:
+                reserved_resources, _ = self.reserved_resources.pop(token)
+                for k in reserved_resources:
+                    for i in reserved_resources[k]:
+                        self.resources[k].append(i)
+            else:
+                self.log_debug(fl_ctx, f"Token {token} is not related to any reserved resources.")
         return None
 
     def allocate_resources(self, resource_requirement: dict, token: str, fl_ctx: FLContext) -> dict:
```

Comment:
both cancel and allocation resources are taken resources from reserved resources. The only difference is one directly return it (use it), another put it back to the general resource pool, It might make sense to write a common function that can be used for both : take_reserved_resource

Reply:
Actually, I found that only two lines are duplicated:

```
if token and token in self.reserved_resources:
    reserved_resources, _ = self.reserved_resources.pop(token)
```

And since the condition is needed, I don't think we need a common function that has only one line.